src/credentials.cpp

```diff
@@ -0,0 +1,254 @@
+		assert(lookup.back() == 0);
+		std::string temp = decrypt(credentials, build_key("*.*", get_system_username()));
+		std::string pass = escape(encrypt(key, build_key(server, login)));
+		size_t pos = temp.find(lookup);
+		if(pos == std::string::npos) {
+			while(!temp.empty() && temp.back() == '\xc') {
+				temp.pop_back();
+			}
+			std::copy(lookup.begin(), lookup.end(), std::back_inserter(temp));
+			std::copy(pass.begin(), pass.end(), std::back_inserter(temp));
+			temp.push_back('\xc');
+			credentials = encrypt(temp, build_key("*.*", get_system_username()));
+			std::fill(temp.begin(), temp.end(), '\0');
+			return;
+		}
+		size_t pos2 = temp.find('\xc', pos + lookup.size() - 1);
+		if(pos2 == std::string::npos) {
+			// TODO: Not sure if this is the right thing to do in this case (or if there's even anything that CAN be done in this case)
+			std::copy(pass.begin(), pass.end(), std::back_inserter(temp));
+			temp.push_back('\xc');
+			credentials = encrypt(temp, build_key("*.*", get_system_username()));
+			std::fill(temp.begin(), temp.end(), '\0');
+			return;
+		}
+		temp.replace(pos + lookup.size(), pos2 - pos - lookup.size(), pass);
+		credentials = encrypt(temp, build_key("*.*", get_system_username()));
+		std::fill(temp.begin(), temp.end(), '\0');
+	}
+
+	void load_credentials()
+	{
+		if(!remember_password()) {
+			return;
+		}
+		// Credentials are in a different file, which is a binary blob
+		filesystem::scoped_istream stream = filesystem::istream_file(filesystem::get_credentials_file(), false);
+		stream->seekg(0, std::ios::end);
+		credentials.clear();
+		credentials.reserve(stream->tellg());
+		stream->seekg(0, std::ios::beg);
+		stream->clear();
+		credentials.assign(std::istreambuf_iterator<char>(*stream), std::istreambuf_iterator<char>());
+	}
+
+	void save_credentials()
+	{
+		if(remember_password()) {
+			try {
+				filesystem::scoped_ostream credentials_file = filesystem::ostream_file(filesystem::get_credentials_file());
+				std::copy(credentials.begin(), credentials.end(), std::ostreambuf_iterator<char>(*credentials_file));
+			} catch(filesystem::io_exception&) {
+				ERR_CFG << "error writing to credentials file '" << filesystem::get_credentials_file() << "'" << std::endl;
+			}
+		} else {
+			filesystem::delete_file(filesystem::get_credentials_file());
+		}
+	}
+}
+
+// TODO: No idea if this is a reasonable way of generating the key.
+std::string build_key(const std::string& server, const std::string& login) {
+	std::ostringstream out;
+	out << '\x2' << login << "\x3wesnoth\x5" << server << '\x7' << get_system_username() << '\xb';
+	return out.str();
+}
+
+// FIXME: XOR encryption is a really terrible choice - swap it out for something better!
+static std::string xor_crypt(std::string text, const std::string& key)
+{
+	const size_t m = key.size();
+	for(size_t i = 0; i < text.size(); i++) {
+		text[i] ^= key[i % m];
+	}
+	return text;
+}
+
+std::string encrypt(const std::string& text, const std::string& key)
```

Comment:
Set a breakpoint here, copy the key value, and your encrypted key is functionally identical to a plain-text key.